### PR TITLE
Transformer model added

### PR DESCRIPTION
**Pull request by preinaj**

### CODE REVIEW

File: ADLStream/models/transformer.py [hunk 1]
```diff
@@ -0,0 +1,550 @@
...
+            )
+
+        output = self.dense(concat_attention)
+
+        return output
+
+
+# Decoder
+class DecoderLayer(tf.keras.layers.Layer):
+    """Class that creates and computes the decoder layer.
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py:141]: Add a white space after this line

File: ADLStream/models/transformer.py [hunk 2]
```diff
@@ -0,0 +1,550 @@
...
+      dff (int): Number of neurons uses in the first layer of the
+        point wise feed forward network.
+      input_size (int): Size of the input.
+      target_size (int): Size of the output.
+      target_shape (tuple): Shape of the output.
+      pe_input (int): Maximum position encoding for the input.
+      pe_target (int): Maximum position encoding for the target.
+      activation (tf.keras.Loss.Activation/String): Activation function for the
+        point wise feed forward network. Defaults to relu.
+      rate (float between 0 and 1): Fraction of the dense units to drop.
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: Maybe "dropout_rate" will be a more descriptive parameter name

File: ADLStream/models/transformer.py [hunk 3]
```diff
@@ -0,0 +1,550 @@
...
+
+    def call(self, inps, training):
+        """
+
+        Args:
+          inps (tuple): Tuple of the input, the mask and the input of the decoder.
+          training (bool): True if the network is in training mode, false if not.
+
+        Returns:
+          tf.Tensor (final_output): Output tensor.
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: Remove the () of the output. Check the [Example Google Style Python Docstrings](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html). 

tf.Tensor: Output tensor.

File: ADLStream/models/transformer.py [hunk 4]
```diff
@@ -0,0 +1,550 @@
...
+        return final_output
+
+    def train_step(self, data):
+        """Function that calculates a train step.
+
+        Args:
+          data (tuple): Tuple with the input and the target tf.Tensors.
+
+        Returns:
+          Dictionary with the model metrics
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: Follow the style for the output:

`dict: training step metrics.`

File: ADLStream/models/transformer.py [hunk 5]
```diff
@@ -0,0 +1,550 @@
...
+        return {m.name: m.result() for m in self.metrics}
+
+    def test_step(self, data):
+        """Function that calculates a train step.
+
+        Args:
+          data (tuple): Tuple of the input and the target tf.Tensors.
+
+        Returns:
+          Dictionary with the model metrics
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: Follow the style for the output:

`dict: training step metrics.`

File: ADLStream/models/transformer.py [hunk 6]
```diff
@@ -0,0 +1,550 @@
...
+                tar, predictions, regularization_losses=self.losses
+            )
+
+        gradients = tape.gradient(loss, self.trainable_variables)
+        self.optimizer.apply_gradients(zip(gradients, self.trainable_variables))
+        self.compiled_metrics.update_state(tar, predictions)
+        return {m.name: m.result() for m in self.metrics}
+
+    def test_step(self, data):
+        """Function that calculates a train step.
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: It should be `test step`, right?

File: ADLStream/models/transformer.py [hunk 7]
```diff
@@ -0,0 +1,550 @@
...
+                output = tf.gather(output, [output.shape[1] - 1], axis=1)
+                tar_inp = tf.concat([tar_inp, output], axis=1)
+        predictions = output
+
+        self.compiled_loss(tar, predictions, regularization_losses=self.losses)
+        self.compiled_metrics.update_state(tar, predictions)
+        return {m.name: m.result() for m in self.metrics}
+
+    def predict_step(self, data):
+        """Function that calculates a train step.
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: Same as before, change `train step` for `prediction step`.

File: ADLStream/models/transformer.py [hunk 8]
```diff
@@ -0,0 +1,550 @@
...
+
+        Args:
+          data (tf.Tensor): Input.
+
+        Returns:
+          tf.Tensor (output): Output tensor.
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: remove `(output)` [ [1]](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

File: ADLStream/models/transformer.py [hunk 9]
```diff
@@ -0,0 +1,550 @@
...
+        num_layers (int): Number of decoder and encoder layers.
+        d_model (int): Number of neurons of the dense layer at the beginning
+          of the encoder and decoder.
+        dff (int): Number of neurons of the rest of dense layers in the model.
+        optimizer (tf.keras.Optimizer): Optimizer that implements theraining algorithm.
+        pe_input (int): Maximum position encoding for the input.
+          Defaults to 1000.
+        pe_output (int): Maximum position encoding for the output.
+          Defaults to 1000.
+        rate (float between 0 and 1): Fraction of the dense units to drop.
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: Similarly to previous comment, `dropout_rate` would be a more descriptive arg name.

File: ADLStream/models/transformer.py [hunk 10]
```diff
@@ -0,0 +1,550 @@
...
+            learning_rate, beta_1=0.9, beta_2=0.98, epsilon=1e-9
+        )
+
+    att = input_shape[-1]
+    inp_len = input_shape[1]
+    inp = np.arange(inp_len * att).reshape((1, inp_len, att))
+    tar_inp = np.arange(output_size).reshape((1, output_shape[0], att))
+
+    model.compile(optimizer=optimizer, loss=loss)
+    model.call((inp, None, tar_inp), False)
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py:552]: Please, add a comment specifying the purpose of this call.

File: ADLStream/models/transformer.py [hunk 11]
```diff
@@ -0,0 +1,550 @@
...
+
+        return tf.math.rsqrt(self.d_model) * tf.math.minimum(arg1, arg2)
+
+
+def Transformer(
+    input_shape,
+    output_size,
+    loss,
+    optimizer,
+    output_shape,
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py:486]: To be consistent with the other models and to avoid errors when the user does not specify the model's parameters, set a default value for all parameter except for the first four (`input_shape`, `output_size`, `loss`, `optimizer`)

File: ADLStream/models/transformer.py [hunk 12]
```diff
@@ -0,0 +1,550 @@
...
+    activation="relu",
+):
+
+    """Transformer
+
+    Args:
+        input_shape (tuple): Shape of the input data.
+        output_size (int): Number of neurons of the last layer.
+        loss (tf.keras.Loss): Loss to be use for training.
+
```
Comment (pedrolarben) [inline ADLStream/models/transformer.py]: remove this empty line

Comment (pedrolarben) [reply ADLStream/models/transformer.py]: And add the description of the `optimizer` parameter. Please specify the meaning of 'custom'.

**Review by pedrolarben — CHANGES_REQUESTED** [review]
Thanks for your contribution! please address the feedback comments before merging.

**Review by pedrolarben — APPROVED** [review]
LGTM
